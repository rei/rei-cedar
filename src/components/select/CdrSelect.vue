<script>
export default {
  inheritAttrs: false,
  customOptions: {}
}
</script>

<script setup>
import { useCssModule, useSlots, useAttrs, computed } from 'vue';
import IconCaretDown from '../icon/comps/caret-down.vue';
import CdrLabelStandalone from '../labelStandalone/CdrLabelStandalone.vue';
import CdrFormError from '../formError/CdrFormError.vue';
import sizeProps from '../../props/size.js';
import backgroundProps from '../../props/background.js';
import mapClasses from '../../utils/mapClasses.js';
import uid from '../../utils/uid.js';
const emit = defineEmits(['update:modelValue'])
const props = defineProps({
  /**
   * `id` for the select that is mapped to the label `for` attribute. If one is not provided, it will be generated.
  */
  id: {
    type: String,
  },
  /**
   * Label text. This is required for a11y even if hiding the label with `hideLabel`.
  */
  label: {
    type: String,
    required: true,
  },
  /**
   * Removes the label element but sets the select `aria-label` to `label` text for a11y.
  */
  hideLabel: {
    type: Boolean,
    default: false,
  },
  /**
   * Adds an option that is disabled and selected by default to serve as a `placeholder` for the select.
  */
  prompt: String,
  /**
   * Build options programatically with data. Array of objects [{ text: String, value: String}] to give greater control. Array of strings ['String'] for simpler setup (value and text will be the same).
  */
  options: {
    type: Array,
  },
  // Set which background type the select renders on
  background: backgroundProps,
  size: sizeProps,
  // Set error styling
  error: {
    type: [Boolean, String],
    default: false,
  },
  /**
  * Override the error message role, default is `status`.
  */
  errorRole: {
    type: String,
    required: false,
    default: 'status',
  },
  modelValue: {
    type: [String, Number, Boolean, Object, Array, Symbol, Function],
  },
  disabled: Boolean,
  required: Boolean,
  optional: Boolean,
  multiple: Boolean,
  multipleSize: Number,
})
const slots = useSlots();
const attrs = useAttrs();
const baseClass = 'cdr-select';
const hasHelper = slots['helper-text'];
const hasInfo = slots.info;
const hasInfoAction = slots['info-action'];
const hasPreIcon = slots['pre-icon'];
const uniqueId = props.id ? props.id : uid();


const multipleClass = computed(() => props.multiple && 'cdr-select--multiple');
const promptClass = computed(() => !props.modelValue && 'cdr-select__prompt');
const preIconClass = computed(() => hasPreIcon && 'cdr-select--preicon');
const errorClass = computed(() => props.error && 'cdr-select--error');
const backgroundClass = computed(() => `cdr-select--${props.background}`);
const sizeClass = computed(() => props.size && `${baseClass}--${props.size}`);
const caretDisabledClass = computed(() => props.disabled && 'cdr-select__caret--disabled');

const describedby = computed(() => {
  return [
    slots['helper-text'] ? `${uniqueId}-helper-text-top` : '',
    attrs['aria-describedby'],
  ].filter((x) => x).join(' ');
})

// TODO: refactor, would be much clearer as a 1-2 liner
const computedOpts = computed(() => {
  const optsArr = [];
  if (props.options) {
    props.options.forEach((o) => {
      const optObj = {};
      let text = '';
      let val = '';
      if (typeof o === 'string') {
        text = o;
        val = o;
      } else {
        const { text: t, value: v } = o;
        text = t;
        val = v;
      }
      optObj.text = text;
      optObj.value = val;
      optsArr.push(optObj);
    });
  }
  return optsArr;
});

const selectModel = computed({
  get() {
    return props.modelValue
  },
  set(newValue) {
    emit('update:modelValue', newValue)
  }
})
const style = useCssModule();
</script>

<template>
  <cdr-label-standalone
    :for-id="uniqueId"
    :label="label"
    :hide-label="hideLabel"
    :required="required"
    :optional="optional"
    :disabled="disabled"
  >
    <template
      #helper
      v-if="hasHelper"
    >
      <slot name="helper-text" />
    </template>
    <template
      #info
      v-if="hasInfo"
    >
      <slot name="info" />
    </template>

    <!-- default slot -->
    <div :class="style['cdr-select-wrap']">
      <span
        v-if="hasPreIcon"
        :class="style['cdr-select__pre-icon']"
      >
        <slot name="pre-icon" />
      </span>

      <select
        :id="uniqueId"
        :class="mapClasses(style,
          baseClass,
          sizeClass,
          promptClass,
          multipleClass,
          backgroundClass,
          errorClass,
          preIconClass,
        )"
        :multiple="multiple"
        :size="multipleSize"
        :disabled="disabled"
        :aria-required="required || null"

        :aria-invalid="!!error || null"
        :aria-errormessage="(!!error && `${uniqueId}-error`) || null"
        v-bind="$attrs"
        :aria-describedby="describedby || null"
        :value="modelValue"
        v-model="selectModel"
      >
        <option
          v-if="prompt"
          :class="style['cdr-select__prompt']"
          value=""
          disabled
        >
          {{ prompt }}
        </option>
        <option
          v-for="option in computedOpts"
          :key="option.text"
          :value="option.value"
        >
          {{ option.text }}
        </option>
        <slot />
      </select>

      <icon-caret-down :class="mapClasses(style, 'cdr-select__caret', caretDisabledClass)" />
    </div>

    <template #info-action v-if="hasInfoAction">
      <slot name="info-action" />
    </template>

    <template #error v-if="error">
      <cdr-form-error
        :error="error"
        :role="errorRole"
        :id="`${uniqueId}-error`"
        v-if="error"
      >
        <template #error>
          <slot name="error" />
        </template>
      </cdr-form-error>
    </template>
  </cdr-label-standalone>
</template>

<style lang="scss" module src="./styles/CdrSelect.module.scss">
</style>
