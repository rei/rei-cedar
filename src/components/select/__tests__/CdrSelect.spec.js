// import { mount } from '../../../../test/vue-jest-style-workaround.js';
import { mount } from '@vue/test-utils';
import CdrSelect from '../CdrSelect.vue';

describe('CdrSelect', () => {
  it('renders correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'Label Test',
        id: 'renders'
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders error state correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'Label Test',
        id: 'renders',
        error: 'What happened?'
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('hide-label applies sr-only to label element', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        hideLabel: true,
        id: 'test',
      },
    });
    expect(wrapper.find('label').classes()).toContain('cdr-label-standalone__label--sr-only');
  });

  it('renders a prompt', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        prompt: 'what is it',
        id: 'test',
      },
    });
    expect(wrapper.find('option').text()).toBe('what is it');
  });

  it('sets select name attribute correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'Label Test',
        id: 'test',
      },
      attrs: {
        name: 'testing',
      }
    });
    expect(wrapper.find('select').attributes('name')).toBe('testing');
  });

  it('sets select disabled attribute correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
      },
      attrs: {
        disabled: true,
      }
    });
    expect(wrapper.find('select').attributes('disabled')).toBe('');
  });

  it('sets select disabled caret classname correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
      },
      attrs: {
        disabled: true,
      }
    });
    const caret = wrapper.find('svg');
    expect(caret.classes()).toContain('cdr-select__caret--disabled');
  });

  it('sets select required attribute correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        required: true,
        id: 'test',
      },
    });
    expect(wrapper.find('select').attributes('aria-required')).toBe('true');
  });

  it('sets select autofocus attribute correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
      },
      attrs: {
        autofocus: true,
      }
    });
    expect(wrapper.find('select').attributes('autofocus')).toBe('');
  });

  it('sets select multiple attribute correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'Label Test',
        multiple: true,
        value: [],
        id: 'test',
      },
    });
    expect(wrapper.find('select').attributes('multiple')).toBe('');
  });

// TODO: something wrong with v-model binding for CdrSelect?
  // xit('emits change event with correct value', async () => {
  //   const wrapper = mount(CdrSelect, {
  //     propsData: {
  //       label: 'test',
  //       value: '4',
  //       options: ['3', '4'],
  //       id: 'test',
  //     },
  //   });
  //   const select = wrapper.find('select');
  //   const options = select.findAll('option');
  //   options[0].setSelected();
  //   await wrapper.vm.$nextTick();
  //   expect(wrapper.emitted().change[0][0]).toBe('3');
  //   expect(wrapper.emitted().change[0][1] instanceof Event).toBeTruthy();
  // });

  // xit('emits change event with correct value for multiple', () => {
  //   const wrapper = mount(CdrSelect, {
  //     propsData: {
  //       label: 'test',
  //       multiple: true,
  //       id: 'test',
  //       value: ['1', '2'],
  //       options: [{
  //         value: '1',
  //         text: 'one',
  //       },
  //       {
  //         value: '2',
  //         text: 'two',
  //       },
  //       {
  //         value: '3',
  //         text: 'three',
  //       }],
  //     },
  //   });
  //   const select = wrapper.find('select');
  //   const options = select.findAll('option');
  //   console.log('multiple opt', options)
  //   options[0].element.selected = true;
  //   options[1].element.selected = false;
  //   options[2].element.selected = true;
  //   select.trigger('change');
  //   console.log('multi', wrapper.emitted())
  //   expect(wrapper.emitted().change[0][0]).toEqual(['1', '3']);
  //   expect(wrapper.emitted().change[0][1] instanceof Event).toBeTruthy();
  // });

  it('updating v-model data updates the select', async () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        value: '4',
        id: 'test',
        options: ['3', '4'],
      },
    });
    const select = wrapper.find('select');
    const options = select.findAll('option');
    wrapper.setProps({modelValue: '3'});
    await wrapper.vm.$nextTick();
    expect(select.element.value).toBe('3');
    expect(options[0].element.selected).toBeTruthy();
  });

  it('renders info action slot', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'info-action'
      },
      slots: {
        'info-action': 'ðŸ¤ ',
      },
    });
    expect(wrapper.find('.cdr-label-standalone__info-action').text()).toBe('ðŸ¤ ');
  });

  it('renders helper-text slot', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'info-action'
      },
      slots: {
        'helper-text': 'yee haw',
      },
    });
    expect(wrapper.find('.cdr-label-standalone__helper').text()).toBe('yee haw');
  });

  it('renders info slot', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'info-action',
      },
      slots: {
        'info': 'howdy',
      },
    });
    expect(wrapper.find('.cdr-label-standalone__info').text()).toBe('howdy');
  });

  it('renders pre-icon slot', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
      },
      slots: {
        'pre-icon': 'ðŸ¤ ',
      },
    });
    expect(wrapper.find('.cdr-select__pre-icon').text()).toBe('ðŸ¤ ');
  });

  it('renders error slot when error state is active', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
        error: true,
      },
      slots: {
        'error': 'whoops',
      },
    });
    expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
  });

  it('renders text when passed as error prop', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
        error: 'wrong!',
      },
    });
    expect(wrapper.find('.cdr-form-error').text()).toBe('wrong!');
  });

  it('does not render error slot when error state is inactive', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'test',
        error: false,
      },
      slots: {
        'error': 'whoops',
      },
    });
    expect(wrapper.find('.cdr-form-error').exists()).toBe(false);
  });

  it('helper text slots are linked to select via aria-describedby', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'aria-test',
      },
      slots: {
        'helper-text': 'extremely helpful'
      },
    });
    expect(wrapper.find('select').attributes('aria-describedby')).toBe('aria-test-helper-text-top');
  });

  it('dynamic aria-describedby is merged with native attr', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'aria-test',
      },
      attrs: {
        'aria-describedby': 'foo',
      },
      slots: {
        'helper-text': 'extremely helpful'
      },
    });
    expect(wrapper.find('select').attributes('aria-describedby')).toBe('aria-test-helper-text-top foo');
  });

  it('does not apply aria-describedby if attr or helper slots are not present', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'test',
        id: 'aria-test',
      },
    });
    expect(wrapper.find('select').attributes('aria-describedby')).toBe(undefined);
  });
});
