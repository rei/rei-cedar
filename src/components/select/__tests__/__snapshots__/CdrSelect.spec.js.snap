// Vitest Snapshot v1

exports[`CdrSelect > renders correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-select-wrap"
    >
      <!--v-if-->
      <select
        class="cdr-select cdr-select__prompt cdr-select--primary"
        id="renders"
      >
        <!--v-if-->
        
        
        
        
      </select>
      <svg
        aria-hidden="true"
        class="cdr-icon cdr-select__caret"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        
        
        <path
          d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 00-1.414-1.414L12 13.586 7.705 9.29a1 1 0 00-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
          role="presentation"
        />
        
        <!--v-if-->
      </svg>
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    
  </div>
</div>
`;

exports[`CdrSelect > renders error state correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-select-wrap"
    >
      <!--v-if-->
      <select
        aria-errormessage="renders-error"
        aria-invalid="true"
        class="cdr-select cdr-select__prompt cdr-select--primary cdr-select--error"
        id="renders"
      >
        <!--v-if-->
        
        
        
        
      </select>
      <svg
        aria-hidden="true"
        class="cdr-icon cdr-select__caret"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        
        
        <path
          d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 00-1.414-1.414L12 13.586 7.705 9.29a1 1 0 00-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
          role="presentation"
        />
        
        <!--v-if-->
      </svg>
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    <div
      class="cdr-form-error"
      id="renders-error"
      role="status"
    >
      <span
        class="cdr-form-error__icon"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-icon--small cdr-icon--inherit-color"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <g>
            <path
              d="M15.293 16.707a1 1 0 001.414-1.414L9.121 7.707a1 1 0 10-1.414 1.414l7.586 7.586z"
              role="presentation"
            />
            <path
              clip-rule="evenodd"
              d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10zm-2 0a8 8 0 11-16 0 8 8 0 0116 0z"
              fill-rule="evenodd"
            />
          </g>
          
          <!--v-if-->
        </svg>
      </span>
      
      <span>
        What happened?
      </span>
      
    </div>
    
  </div>
</div>
`;
